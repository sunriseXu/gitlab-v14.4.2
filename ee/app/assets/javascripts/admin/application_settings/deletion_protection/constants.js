import { __, s__ } from '~/locale';

export const I18N_DELETION_PROTECTION = {
  heading: s__('DeletionSettings|Deletion protection'),
  helpText: s__(
    'DeletionSettings|Retention period that deleted groups and projects will remain restorable. Personal projects are always deleted immediately. Some groups can opt-out their projects.',
  ),
  learnMore: __('Learn more.'),
  keepDeleted: s__('DeletionSettings|Keep deleted'),
  deleteImmediately: s__('DeletionSettings|None, delete immediately'),
  for: __('for'),
  days: __('days'),
  groupsOnly: __('groups only'),
  groupsAndProjects: __('groups and projects'),
};
