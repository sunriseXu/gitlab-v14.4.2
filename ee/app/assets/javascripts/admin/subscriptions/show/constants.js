import { __, s__ } from '~/locale';
import { PROMO_URL } from 'jh_else_ce/lib/utils/url_utility';

export const subscriptionMainTitle = s__('SuperSonics|Your subscription');
export const subscriptionActivationNotificationText = s__(
  `SuperSonics|Your subscription was successfully activated. You can see the details below.`,
);
export const subscriptionActivationFutureDatedNotificationTitle = s__(
  'SuperSonics|Your future dated license was successfully added',
);
export const subscriptionActivationFutureDatedNotificationMessage = s__(
  'SuperSonics|You have successfully added a license that activates on %{date}. Please see the subscription history table below for more details.',
);
export const subscriptionActivationInsertCode = __(
  "If you've purchased or renewed your subscription and have an activation code, please enter it below to start the activation process.",
);
export const howToActivateSubscription = s__(
  'SuperSonics|Learn how to %{linkStart}activate your subscription%{linkEnd}.',
);
export const subscriptionHistoryFailedTitle = s__('SuperSonics|Subscription unavailable');
export const subscriptionHistoryFailedMessage = s__(
  'SuperSonics|Your %{subscriptionEntryName} cannot be displayed at the moment. Please refresh the page to try again.',
);
export const currentSubscriptionsEntryName = s__('SuperSonics|current subscription');
export const pastSubscriptionsEntryName = s__('SuperSonics|past subscriptions');
export const futureSubscriptionsEntryName = s__('SuperSonics|future subscriptions');

export const cancelLabel = __('Cancel');
export const activateLabel = s__('AdminUsers|Activate');
export const activateSubscription = s__('SuperSonics|Activate subscription');
export const activateCloudLicense = s__('SuperSonics|Enter activation code');
export const noActiveSubscription = s__(`SuperSonics|You do not have an active subscription`);
export const subscriptionDetailsHeaderText = s__('SuperSonics|Subscription details');
export const licensedToHeaderText = s__('SuperSonics|Licensed to');
export const manageSubscriptionButtonText = s__('SuperSonics|Manage');
export const syncSubscriptionButtonText = s__('SuperSonics|Sync subscription details');
export const copySubscriptionIdButtonText = __('Copy');
export const licenseFileText = __('Legacy license');
export const onlineCloudLicenseText = s__('SuperSonics|Online license');
export const offlineCloudLicenseText = s__('SuperSonics|Offline license');
export const detailsLabels = {
  address: __('Address'),
  company: __('Company'),
  email: __('Email'),
  id: __('ID'),
  lastSync: __('Last Sync'),
  name: licensedToHeaderText,
  plan: __('Plan'),
  type: __('Type'),
  expiresAt: __('Renews'),
  startsAt: __('Started'),
};

export const subscriptionTable = {
  activatedAt: __('Activated on'),
  expiresOn: __('Expires on'),
  seats: __('Seats'),
  startsAt: __('Valid From'),
  title: __('Subscription History'),
  type: __('Type'),
};

export const subscriptionActivationForm = {
  activationCode: s__('SuperSonics|Activation code'),
  activationCodeFeedback: s__(
    'SuperSonics|The activation code should be a 24-character alphanumeric string',
  ),
  pasteActivationCode: s__('SuperSonics|Paste your activation code'),
  acceptTerms: s__(
    'SuperSonics|I agree that my use of the GitLab Software is subject to the Subscription Agreement located at the %{linkStart}Terms of Service%{linkEnd}, unless otherwise agreed to in writing with GitLab.',
  ),
  acceptTermsFeedback: s__('SuperSonics|Please agree to the Subscription Agreement'),
};

export const subscriptionSyncStatus = {
  SYNC_FAILURE: 'SYNC_FAILURE',
  SYNC_PENDING: 'SYNC_PENDING',
};

export const subscriptionTypes = {
  ONLINE_CLOUD: 'online_cloud',
  OFFLINE_CLOUD: 'offline_cloud',
  LEGACY_LICENSE: 'legacy_license',
};

export const trialCard = {
  title: s__('SuperSonics|Free trial'),
  description: s__(
    'SuperSonics|You can start a free trial of GitLab Ultimate without any obligation or payment details.',
  ),
  startTrial: s__('SuperSonics|Start free trial'),
};

export const buySubscriptionCard = {
  title: __('Subscription'),
  description: s__(
    'SuperSonics|Ready to get started? A GitLab plan is ideal for scaling organizations and for multi team usage.',
  ),
  buttonLabel: s__('SuperSonics|Buy subscription'),
};

export const SUBSCRIPTION_ACTIVATION_FAILURE_EVENT = 'subscription-activation-failure';
export const SUBSCRIPTION_ACTIVATION_SUCCESS_EVENT = 'subscription-activation-success';
export const SUBSCRIPTION_ACTIVATION_FINALIZED_EVENT = 'subscription-activation-finalized';

// Server-sent error messages (or regexes to match those messages) to detect error type
export const INVALID_ACTIVATION_CODE_SERVER_ERROR = 'invalid activation code';
export const EXPIRED_LICENSE_SERVER_ERROR = 'This license has already expired.';
export const SUBSCRIPTION_NOT_FOUND_SERVER_ERROR = 'without cloud compatible subscription';
export const SUBSCRIPTION_OVERAGES_SERVER_ERROR_REGEX = /(This GitLab installation currently has|During the year before this license started, this GitLab installation had) (\d+) active users?, exceeding this license's limit of (\d+) by (\d+) users?\. Please add a license for at least (\d+) users? or contact sales at https:\/\/about\.gitlab\.com\/sales\//;
export const SUBSCRIPTION_INSUFFICIENT_TRUE_UP_SERVER_ERROR_REGEX = /You have applied a True-up for (\d+) users? but you need one for (\d+) users?\. Please contact sales at https:\/\/about\.gitlab\.com\/sales\//;

export const CONNECTIVITY_ERROR = 'CONNECTIVITY_ERROR';
export const INVALID_CODE_ERROR = 'INVALID_CODE_ERROR';

export const supportLink = `${PROMO_URL}/support/#contact-support`;
export const subscriptionBannerTitle = s__('SuperSonics|Cloud licensing');
export const subscriptionBannerText = s__(
  "SuperSonics|Cloud licensing is now available. It's an easier way to activate instances and manage subscriptions. Read more about it in our %{blogPostLinkStart}blog post%{blogPostLinkEnd}. Activation codes are available in the %{portalLinkStart}Customers Portal%{portalLinkEnd}.",
);
export const subscriptionBannerBlogPostUrl =
  'https://about.gitlab.com/blog/2021/07/20/improved-billing-and-subscription-management/';
export const exportLicenseUsageBtnText = s__('SuperSonics|Export license usage file');

export const instanceHasFutureLicenseBanner = {
  title: s__('SuperSonics|You have a future dated license'),
  message: s__(
    'SuperSonics|You have added a license that activates on %{date}. Please see the subscription history table below for more details.',
  ),
};
