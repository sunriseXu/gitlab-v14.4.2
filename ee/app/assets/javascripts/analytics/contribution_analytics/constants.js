export const CHART_HEIGHT = 350;
export const INNER_CHART_HEIGHT = 200;
export const CHART_X_AXIS_ROTATE = 45;
export const CHART_X_AXIS_NAME_TOP_PADDING = 55;
