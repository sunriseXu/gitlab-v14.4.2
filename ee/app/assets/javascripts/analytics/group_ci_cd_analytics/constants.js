export const STAT_ERROR_PLACEHOLDER = '-';
export const formatWithUtc = true;
