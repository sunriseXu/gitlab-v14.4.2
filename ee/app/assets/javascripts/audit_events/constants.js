import { n__, s__, __ } from '~/locale';
import { OPERATOR_IS_ONLY } from '~/vue_shared/components/filtered_search_bar/constants';
import GroupToken from './components/tokens/group_token.vue';
import MemberToken from './components/tokens/member_token.vue';
import ProjectToken from './components/tokens/project_token.vue';
import UserToken from './components/tokens/user_token.vue';

const DEFAULT_TOKEN_OPTIONS = {
  operators: OPERATOR_IS_ONLY,
  unique: true,
};

// Due to the i18n eslint rule we can't have a capitalized string even if it is a case-aware URL param
/* eslint-disable @gitlab/require-i18n-strings */
export const ENTITY_TYPES = {
  USER: 'User',
  AUTHOR: 'Author',
  GROUP: 'Group',
  PROJECT: 'Project',
};
/* eslint-enable @gitlab/require-i18n-strings */

export const AUDIT_FILTER_CONFIGS = [
  {
    ...DEFAULT_TOKEN_OPTIONS,
    icon: 'user',
    title: s__('AuditLogs|User Events'),
    type: 'user',
    entityType: ENTITY_TYPES.USER,
    token: UserToken,
  },
  {
    ...DEFAULT_TOKEN_OPTIONS,
    icon: 'user',
    title: s__('AuditLogs|Member Events'),
    type: 'member',
    entityType: ENTITY_TYPES.AUTHOR,
    token: MemberToken,
  },
  {
    ...DEFAULT_TOKEN_OPTIONS,
    icon: 'bookmark',
    title: s__('AuditLogs|Project Events'),
    type: 'project',
    entityType: ENTITY_TYPES.PROJECT,
    token: ProjectToken,
  },
  {
    ...DEFAULT_TOKEN_OPTIONS,
    icon: 'group',
    title: s__('AuditLogs|Group Events'),
    type: 'group',
    entityType: ENTITY_TYPES.GROUP,
    token: GroupToken,
  },
];

export const AVAILABLE_TOKEN_TYPES = AUDIT_FILTER_CONFIGS.map((token) => token.type);

export const MAX_DATE_RANGE = 31;

export const SAME_DAY_OFFSET = 1;

// This creates a date with zero time, making it simpler to match to the query date values
export const CURRENT_DATE = new Date(new Date().toDateString());

export const AUDIT_EVENTS_TAB_TITLES = {
  LOG: s__('AuditLogs|Log'),
  STREAM: s__('AuditStreams|Streams'),
};

export const streamsLabel = (count) =>
  n__('AuditStreams|%d destination', 'AuditStreams|%d destinations', count);
export const ADD_STREAM = s__('AuditStreams|Add streaming destination');
export const ADD_STREAM_MESSAGE = s__('AuditStreams|Stream added successfully');
export const UPDATE_STREAM_MESSAGE = s__('AuditStreams|Stream updated successfully');
export const DELETE_STREAM_MESSAGE = s__('AuditStreams|Stream deleted successfully');

export const STREAM_ITEMS_I18N = {
  VERIFICATION_TOKEN_TOOLTIP: s__('AuditStreams|Verification token'),
  EDIT_BUTTON_LABEL: s__('AuditStreams|Edit %{link}'),
  DELETE_BUTTON_LABEL: s__('AuditStreams|Delete %{link}'),
  EDIT_BUTTON_TOOLTIP: __('Edit'),
  DELETE_BUTTON_TOOLTIP: __('Delete'),
};

export const ADD_STREAM_EDITOR_I18N = {
  WARNING_TITLE: s__('AuditStreams|Destinations receive all audit event data'),
  WARNING_CONTENT: s__(
    'AuditStreams|This could include sensitive information. Make sure you trust the destination endpoint.',
  ),
  DESTINATION_URL_LABEL: s__('AuditStreams|Destination URL'),
  DESTINATION_URL_PLACEHOLDER: 'https://api.gitlab.com',
  HEADERS_LABEL: s__('AuditStreams|Custom HTTP headers (optional)'),
  TABLE_COLUMN_NAME_LABEL: s__('AuditStreams|Header'),
  TABLE_COLUMN_VALUE_LABEL: s__('AuditStreams|Value'),
  TABLE_COLUMN_ACTIVE_LABEL: s__('AuditStreams|Active'),
  HEADER_INPUT_PLACEHOLDER: s__('AuditStreams|Add a custom header'),
  HEADER_INPUT_DUPLICATE_ERROR: s__('AuditStreams|A header with this name already exists.'),
  VALUE_INPUT_PLACEHOLDER: s__('AuditStreams|Add a custom value'),
  ADD_HEADER_ROW_BUTTON_TEXT: s__('AuditStreams|Add header'),
  ADD_HEADER_ROW_BUTTON_NAME: s__('AuditStreams|Add another custom header'),
  MAXIMUM_HEADERS_TEXT: s__('AuditStreams|Maximum of %{number} HTTP headers has been reached.'),
  REMOVE_BUTTON_LABEL: s__('AuditStreams|Remove custom header'),
  REMOVE_BUTTON_TOOLTIP: __('Remove'),
  ADD_BUTTON_TEXT: __('Add'),
  ADD_BUTTON_NAME: s__('AuditStreams|Add external stream destination'),
  SAVE_BUTTON_TEXT: __('Save'),
  SAVE_BUTTON_NAME: s__('AuditStreams|Save external stream destination'),
  CANCEL_BUTTON_TEXT: __('Cancel'),
  CANCEL_BUTTON_NAME: s__('AuditStreams|Cancel editing'),
};

export const AUDIT_STREAMS_EMPTY_STATE_I18N = {
  TITLE: s__('AuditStreams|Setup streaming for audit events'),
  DESCRIPTION_1: s__(
    'AuditStreams|Add an HTTP endpoint to manage audit logs in third-party systems.',
  ),
  DESCRIPTION_2: s__('AuditStreams|This is great for keeping everything one place.'),
};

export const AUDIT_STREAMS_NETWORK_ERRORS = {
  FETCHING_ERROR: s__(
    'AuditStreams|An error occurred when fetching external audit event streams. Please try it again',
  ),
  CREATING_ERROR: s__(
    'AuditStreams|An error occurred when creating external audit event stream destination. Please try it again.',
  ),
  UPDATING_ERROR: s__(
    'AuditStreams|An error occurred when updating external audit event stream destination. Please try it again.',
  ),
  DELETING_ERROR: s__(
    'AuditStreams|An error occurred when deleting external audit event stream destination. Please try it again.',
  ),
};

export const MAX_HEADERS = 20;
export const createBlankHeader = () => ({
  id: null,
  name: '',
  value: '',
  active: true,
  disabled: false,
  deletionDisabled: false,
  validationErrors: { name: '' },
});
