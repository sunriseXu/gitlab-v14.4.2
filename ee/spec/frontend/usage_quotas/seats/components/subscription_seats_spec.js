import {
  GlPagination,
  GlButton,
  GlTable,
  GlAvatarLink,
  GlAvatarLabeled,
  GlBadge,
  GlModal,
} from '@gitlab/ui';
import { mount, shallowMount } from '@vue/test-utils';
import Vue from 'vue';
import Vuex from 'vuex';
import StatisticsCard from 'ee/usage_quotas/components/statistics_card.vue';
import StatisticsSeatsCard from 'ee/usage_quotas/components/statistics_seats_card.vue';
import SubscriptionUpgradeInfoCard from 'ee/usage_quotas/seats/components/subscription_upgrade_info_card.vue';
import SubscriptionSeats from 'ee/usage_quotas/seats/components/subscription_seats.vue';

import { CANNOT_REMOVE_BILLABLE_MEMBER_MODAL_CONTENT } from 'ee/usage_quotas/seats/constants';

import { mockDataSeats, mockTableItems } from 'ee_jest/usage_quotas/seats/mock_data';
import { extendedWrapper } from 'helpers/vue_test_utils_helper';
import FilterSortContainerRoot from '~/vue_shared/components/filtered_search_bar/filtered_search_bar_root.vue';

Vue.use(Vuex);

const actionSpies = {
  fetchBillableMembersList: jest.fn(),
  fetchGitlabSubscription: jest.fn(),
  setBillableMemberToRemove: jest.fn(),
  setSearchQuery: jest.fn(),
  changeMembershipState: jest.fn(),
};

const providedFields = {
  namespaceName: 'Test Group Name',
  namespaceId: '1000',
  seatUsageExportPath: '/groups/test_group/-/seat_usage.csv',
  maxFreeNamespaceSeats: 5,
  explorePlansPath: '/groups/test_group/-/billings',
  hasNoSubscription: false,
  hasLimitedFreePlan: false,
  hasReachedFreePlanLimit: false,
  previewFreeUserCap: false,
  activeTrial: false,
};

const fakeStore = ({ initialState, initialGetters }) =>
  new Vuex.Store({
    actions: actionSpies,
    getters: {
      tableItems: () => mockTableItems,
      isLoading: () => false,
      ...initialGetters,
    },
    state: {
      hasError: false,
      namespaceId: 1,
      members: [...mockDataSeats.data],
      total: 300,
      page: 1,
      perPage: 5,
      sort: 'last_activity_on_desc',
      ...providedFields,
      ...initialState,
    },
  });

describe('Subscription Seats', () => {
  let wrapper;

  const createComponent = ({
    initialState = {},
    mountFn = shallowMount,
    initialGetters = {},
    provide = {},
  } = {}) => {
    return extendedWrapper(
      mountFn(SubscriptionSeats, {
        store: fakeStore({ initialState, initialGetters }),
        provide,
      }),
    );
  };

  const findTable = () => wrapper.findComponent(GlTable);

  const findExportButton = () => wrapper.findByTestId('export-button');

  const findSearchBox = () => wrapper.findComponent(FilterSortContainerRoot);
  const findPagination = () => wrapper.findComponent(GlPagination);

  const findAllRemoveUserItems = () => wrapper.findAllByTestId('remove-user');
  const findErrorModal = () => wrapper.findComponent(GlModal);
  const findStatisticsCard = () => wrapper.findComponent(StatisticsCard);
  const findStatisticsSeatsCard = () => wrapper.findComponent(StatisticsSeatsCard);
  const findSubscriptionUpgradeCard = () => wrapper.findComponent(SubscriptionUpgradeInfoCard);
  const findSkeletonLoaderCards = () => wrapper.findByTestId('skeleton-loader-cards');

  const serializeUser = (rowWrapper) => {
    const avatarLink = rowWrapper.findComponent(GlAvatarLink);
    const avatarLabeled = rowWrapper.findComponent(GlAvatarLabeled);

    return {
      avatarLink: {
        href: avatarLink.attributes('href'),
        alt: avatarLink.attributes('alt'),
      },
      avatarLabeled: {
        src: avatarLabeled.attributes('src'),
        size: avatarLabeled.attributes('size'),
        text: avatarLabeled.text(),
      },
    };
  };

  const serializeTableRow = (rowWrapper) => {
    const emailWrapper = rowWrapper.find('[data-testid="email"]');

    return {
      user: serializeUser(rowWrapper),
      email: emailWrapper.text(),
      tooltip: emailWrapper.find('span').attributes('title'),
      removeUserButtonExists: rowWrapper.findComponent(GlButton).exists(),
    };
  };

  const findSerializedTable = (tableWrapper) => {
    return tableWrapper.findAll('tbody tr').wrappers.map(serializeTableRow);
  };

  afterEach(() => {
    wrapper.destroy();
  });

  describe('actions', () => {
    beforeEach(() => {
      wrapper = createComponent();
    });

    it('correct actions are called on create', () => {
      expect(actionSpies.fetchBillableMembersList).toHaveBeenCalled();
    });
  });

  describe('renders', () => {
    beforeEach(() => {
      wrapper = createComponent({
        mountFn: mount,
        initialGetters: {
          tableItems: () => mockTableItems,
        },
      });
    });

    describe('export button', () => {
      it('has the correct href', () => {
        expect(findExportButton().attributes().href).toBe(providedFields.seatUsageExportPath);
      });
    });

    describe('table content', () => {
      it('renders the correct data', () => {
        const serializedTable = findSerializedTable(findTable());

        expect(serializedTable).toMatchSnapshot();
      });
    });

    it('pagination is rendered and passed correct values', () => {
      const pagination = findPagination();

      expect(pagination.props()).toMatchObject({
        perPage: 5,
        totalItems: 300,
      });
    });

    describe('with error modal', () => {
      it('does not render the model if the user is not removable', async () => {
        await findAllRemoveUserItems().at(0).trigger('click');

        expect(findErrorModal().html()).toBe('');
      });

      it('renders the error modal if the user is removable', async () => {
        await findAllRemoveUserItems().at(2).trigger('click');

        expect(findErrorModal().text()).toContain(CANNOT_REMOVE_BILLABLE_MEMBER_MODAL_CONTENT);
      });
    });

    describe('members avatar', () => {
      it('shows the correct avatarLinks length', () => {
        const avatarLinks = findTable().findAllComponents(GlAvatarLink);
        expect(avatarLinks.length).toBe(6);
      });

      it.each(['group_invite', 'project_invite'])(
        'shows the correct badge for membership_type %s',
        (membershipType) => {
          const avatarLinks = findTable().findAllComponents(GlAvatarLink);
          const badgeText = (
            membershipType.charAt(0).toUpperCase() + membershipType.slice(1)
          ).replace('_', ' ');

          avatarLinks.wrappers.forEach((avatarLinkWrapper) => {
            const currentMember = mockTableItems.find(
              (item) => item.user.name === avatarLinkWrapper.attributes().alt,
            );

            if (membershipType === currentMember.user.membership_type) {
              expect(avatarLinkWrapper.findComponent(GlBadge).text()).toBe(badgeText);
            }
          });
        },
      );
    });

    describe('members details', () => {
      it.each`
        membershipType      | shouldShowDetails
        ${'project_invite'} | ${false}
        ${'group_invite'}   | ${false}
        ${'project_member'} | ${true}
        ${'group_member'}   | ${true}
      `(
        'when membershipType is $membershipType, shouldShowDetails should be $shouldShowDetails',
        ({ membershipType, shouldShowDetails }) => {
          const seatCells = findTable().findAll('[data-testid*="seat-cell-"]');

          seatCells.wrappers.forEach((seatCellWrapper) => {
            const currentMember = mockTableItems.find(
              (item) => seatCellWrapper.attributes('data-testid') === `seat-cell-${item.user.id}`,
            );

            if (membershipType === currentMember.user.membership_type) {
              expect(
                seatCellWrapper.find('[data-testid="toggle-seat-usage-details"]').exists(),
              ).toBe(shouldShowDetails);
            }
          });
        },
      );
    });
  });

  describe('statistics cards', () => {
    const defaultInitialState = {
      seatsInSubscription: 3,
      total: 10,
      seatsInUse: 2,
      maxSeatsUsed: 3,
      seatsOwed: 1,
    };

    const defaultProps = {
      helpLink: '/help/subscriptions/gitlab_com/index#how-seat-usage-is-determined',
      totalUnit: null,
      usageUnit: null,
    };

    beforeEach(() => {
      wrapper = createComponent({
        initialState: defaultInitialState,
      });
    });

    it('calls the correct action on create', () => {
      expect(actionSpies.fetchGitlabSubscription).toHaveBeenCalled();
    });

    describe('renders <statistics-card>', () => {
      describe('when group has a subscription', () => {
        it('renders <statistics-card> with the necessary props', () => {
          const statisticsCard = findStatisticsCard();

          expect(statisticsCard.exists()).toBe(true);
          expect(statisticsCard.props()).toMatchObject({
            ...defaultProps,
            description: 'Seats in use / Seats in subscription',
            percentage: 67,
            totalValue: '3',
            usageValue: '2',
            helpTooltip: null,
          });
        });
      });

      describe('when group has no subscription', () => {
        describe('when not on limited free plan', () => {
          beforeEach(() => {
            wrapper = createComponent({
              initialState: {
                ...defaultInitialState,
                hasNoSubscription: true,
                hasLimitedFreePlan: false,
                activeTrial: false,
              },
            });
          });

          it('renders <statistics-card> with the necessary props', () => {
            const statisticsCard = findStatisticsCard();

            expect(statisticsCard.exists()).toBe(true);
            expect(statisticsCard.props()).toMatchObject({
              ...defaultProps,
              description: 'Seats in use / Seats in subscription',
              percentage: 0,
              totalValue: 'Unlimited',
              usageValue: '10',
              helpTooltip: null,
            });
          });

          describe('when on trial', () => {
            beforeEach(() => {
              wrapper = createComponent({
                initialState: {
                  ...defaultInitialState,
                  hasNoSubscription: true,
                  hasLimitedFreePlan: false,
                  activeTrial: true,
                },
              });
            });

            it('renders <statistics-card> with the necessary props', () => {
              const statisticsCard = findStatisticsCard();

              expect(statisticsCard.exists()).toBe(true);
              expect(statisticsCard.props()).toMatchObject({
                ...defaultProps,
                description: 'Seats in use / Seats in subscription',
                percentage: 0,
                totalValue: 'Unlimited',
                usageValue: '10',
                helpTooltip: null,
              });
            });
          });
        });

        describe('when on limited free plan', () => {
          beforeEach(() => {
            wrapper = createComponent({
              initialState: {
                ...defaultInitialState,
                hasNoSubscription: true,
                hasLimitedFreePlan: true,
                activeTrial: false,
              },
            });
          });

          it('renders <statistics-card> with the necessary props', () => {
            const statisticsCard = findStatisticsCard();

            expect(statisticsCard.exists()).toBe(true);
            expect(statisticsCard.props()).toMatchObject({
              ...defaultProps,
              description: 'Seats in use / Seats available',
              percentage: 40,
              totalValue: '5',
              usageValue: '2',
              helpTooltip: 'Free groups are limited to 5 seats.',
            });
          });

          describe('when on trial', () => {
            beforeEach(() => {
              wrapper = createComponent({
                initialState: {
                  ...defaultInitialState,
                  hasNoSubscription: true,
                  hasLimitedFreePlan: true,
                  activeTrial: true,
                },
              });
            });

            it('renders <statistics-card> with the necessary props', () => {
              const statisticsCard = findStatisticsCard();

              expect(statisticsCard.exists()).toBe(true);
              expect(statisticsCard.props()).toMatchObject({
                ...defaultProps,
                description: 'Seats in use / Seats available',
                percentage: 0,
                totalValue: 'Unlimited',
                usageValue: '2',
                helpTooltip:
                  'Free tier and trial groups can invite a maximum of 20 members per day.',
              });
            });
          });
        });

        describe('when preview free user cap', () => {
          beforeEach(() => {
            wrapper = createComponent({
              initialState: {
                ...defaultInitialState,
                hasNoSubscription: true,
                hasLimitedFreePlan: false,
                previewFreeUserCap: true,
              },
            });
          });

          it('renders <statistics-card> with the necessary props', () => {
            const statisticsCard = findStatisticsCard();

            expect(statisticsCard.exists()).toBe(true);
            expect(statisticsCard.props()).toMatchObject({
              ...defaultProps,
              description: 'Seats in use / Seats available',
              percentage: 0,
              totalValue: 'Unlimited',
              usageValue: '2',
            });
          });
        });
      });
    });

    it('renders <statistics-seats-card> with the necessary props', () => {
      const statisticsSeatsCard = findStatisticsSeatsCard();

      expect(findSubscriptionUpgradeCard().exists()).toBe(false);
      expect(statisticsSeatsCard.exists()).toBe(true);
      expect(statisticsSeatsCard.props()).toMatchObject({
        seatsOwed: 1,
        seatsUsed: 3,
      });
    });

    describe('for free namespace with limit', () => {
      beforeEach(() => {
        wrapper = createComponent({
          initialState: { hasNoSubscription: true, hasLimitedFreePlan: true },
        });
      });

      it('renders <subscription-upgrade-info-card> with the necessary props', () => {
        const upgradeInfoCard = findSubscriptionUpgradeCard();

        expect(findStatisticsSeatsCard().exists()).toBe(false);
        expect(upgradeInfoCard.exists()).toBe(true);
        expect(upgradeInfoCard.props()).toMatchObject({
          maxNamespaceSeats: providedFields.maxFreeNamespaceSeats,
          explorePlansPath: providedFields.explorePlansPath,
          activeTrial: false,
        });
      });
    });

    describe('for free namespace with free user cap preview enabled', () => {
      beforeEach(() => {
        wrapper = createComponent({
          initialState: {
            hasNoSubscription: true,
            hasLimitedFreePlan: false,
            previewFreeUserCap: true,
          },
        });
      });

      it('renders <subscription-upgrade-info-card> with the necessary props', () => {
        const upgradeInfoCard = findSubscriptionUpgradeCard();

        expect(findStatisticsSeatsCard().exists()).toBe(false);
        expect(upgradeInfoCard.exists()).toBe(true);
        expect(upgradeInfoCard.props()).toMatchObject({
          maxNamespaceSeats: providedFields.maxFreeNamespaceSeats,
          explorePlansPath: providedFields.explorePlansPath,
          activeTrial: false,
        });
      });
    });
  });

  describe('Loading state', () => {
    describe('When nothing is loading', () => {
      beforeEach(() => {
        wrapper = createComponent();
      });

      it('displays the table in a non-busy state', () => {
        expect(findTable().attributes('busy')).toBe(undefined);
      });
    });

    describe.each([
      [true, false],
      [false, true],
    ])('Busy when isLoading=%s and hasError=%s', (isLoading, hasError) => {
      beforeEach(() => {
        wrapper = createComponent({
          initialGetters: { isLoading: () => isLoading },
          initialState: { hasError },
        });
      });

      it('displays loading skeletons instead of statistics cards', () => {
        expect(findSkeletonLoaderCards().exists()).toBe(true);
        expect(findStatisticsCard().exists()).toBe(false);
        expect(findStatisticsSeatsCard().exists()).toBe(false);
      });

      it('displays table in busy state', () => {
        expect(findTable().attributes('busy')).toBe('true');
      });
    });
  });

  describe('search box', () => {
    beforeEach(() => {
      wrapper = createComponent();
    });

    it('input event triggers the setSearchQuery action', async () => {
      const SEARCH_STRING = 'search string';

      // fetchBillableMembersList is called once on created()
      expect(actionSpies.fetchBillableMembersList).toHaveBeenCalledTimes(1);

      await findSearchBox().vm.$emit('onFilter', [
        { type: 'filtered-search-term', value: { data: SEARCH_STRING } },
      ]);

      expect(actionSpies.setSearchQuery).toHaveBeenCalledWith(expect.any(Object), SEARCH_STRING);
    });
  });

  describe('pending members alert', () => {
    it.each`
      pendingMembersPagePath | pendingMembersCount | hasLimitedFreePlan | previewFreeUserCap | shouldBeRendered
      ${undefined}           | ${undefined}        | ${false}           | ${false}           | ${false}
      ${undefined}           | ${0}                | ${false}           | ${false}           | ${false}
      ${'fake-path'}         | ${0}                | ${false}           | ${false}           | ${false}
      ${'fake-path'}         | ${3}                | ${true}            | ${false}           | ${false}
      ${'fake-path'}         | ${3}                | ${false}           | ${true}            | ${false}
      ${'fake-path'}         | ${3}                | ${false}           | ${false}           | ${true}
    `(
      'rendering alert is $shouldBeRendered when pendingMembersPagePath=$pendingMembersPagePath and pendingMembersCount=$pendingMembersCount and hasLimitedFreePlan=$hasLimitedFreePlan and previewFreeUserCap=$previewFreeUserCap',
      ({
        pendingMembersPagePath,
        pendingMembersCount,
        shouldBeRendered,
        hasLimitedFreePlan,
        previewFreeUserCap,
      }) => {
        wrapper = createComponent({
          initialState: {
            pendingMembersCount,
            pendingMembersPagePath,
            hasLimitedFreePlan,
            previewFreeUserCap,
          },
        });

        expect(wrapper.find('[data-testid="pending-members-alert"]').exists()).toBe(
          shouldBeRendered,
        );
      },
    );
  });
});
