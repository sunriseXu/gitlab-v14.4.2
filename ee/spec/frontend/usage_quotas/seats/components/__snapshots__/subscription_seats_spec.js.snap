// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subscription Seats renders table content renders the correct data 1`] = `
Array [
  Object {
    "email": "administrator@email.com",
    "removeUserButtonExists": true,
    "tooltip": undefined,
    "user": Object {
      "avatarLabeled": Object {
        "size": "32",
        "src": "path/to/img_administrator",
        "text": "Administrator   @root",
      },
      "avatarLink": Object {
        "alt": "Administrator",
        "href": "path/to/administrator",
      },
    },
  },
  Object {
    "email": "agustin_walker@email.com",
    "removeUserButtonExists": true,
    "tooltip": undefined,
    "user": Object {
      "avatarLabeled": Object {
        "size": "32",
        "src": "path/to/img_agustin_walker",
        "text": "Agustin Walker   @lester.orn",
      },
      "avatarLink": Object {
        "alt": "Agustin Walker",
        "href": "path/to/agustin_walker",
      },
    },
  },
  Object {
    "email": "Private",
    "removeUserButtonExists": true,
    "tooltip": "An email address is only visible for users with public emails.",
    "user": Object {
      "avatarLabeled": Object {
        "size": "32",
        "src": "path/to/img_joella_miller",
        "text": "Joella Miller  
                Group invite
                @era",
      },
      "avatarLink": Object {
        "alt": "Joella Miller",
        "href": "path/to/joella_miller",
      },
    },
  },
  Object {
    "email": "jdoe@email.com",
    "removeUserButtonExists": true,
    "tooltip": undefined,
    "user": Object {
      "avatarLabeled": Object {
        "size": "32",
        "src": "path/to/img_john_doe",
        "text": "John Doe   
                Project invite
               @jdoe",
      },
      "avatarLink": Object {
        "alt": "John Doe",
        "href": "path/to/john_doe",
      },
    },
  },
  Object {
    "email": "jsnow@email.com",
    "removeUserButtonExists": true,
    "tooltip": undefined,
    "user": Object {
      "avatarLabeled": Object {
        "size": "32",
        "src": "path/to/img_john_snow",
        "text": "John Snow   @jown.snow",
      },
      "avatarLink": Object {
        "alt": "John Snow",
        "href": "path/to/john_snow",
      },
    },
  },
  Object {
    "email": "current_user@email.com",
    "removeUserButtonExists": true,
    "tooltip": undefined,
    "user": Object {
      "avatarLabeled": Object {
        "size": "32",
        "src": "path/to/img_curent_user",
        "text": "Current user   @current.user",
      },
      "avatarLink": Object {
        "alt": "Current user",
        "href": "path/to/current_user",
      },
    },
  },
]
`;
